File: smalivm/src/test/java/org/cf/smalivm/opcode/AGetOpTest.java
package org.cf.smalivm.opcode;

import org.cf.smalivm.VMState;
import org.cf.smalivm.VMTester;
import org.cf.smalivm.context.ExecutionGraph;
import org.cf.smalivm.context.HeapItem;
import org.cf.smalivm.context.MethodState;
import org.cf.smalivm.type.UnknownValue;
import org.cf.util.ClassNameUtils;
import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;

import static org.junit.Assert.assertFalse;

public class AGetOpTest {

    private static final String CLASS_NAME = "Laget_test;";

    private VMState expected;
    private VMState initial;

    private static void testException(String methodDescriptor, Class<?> exceptionClass, VMState initial) {
        ExecutionGraph graph = VMTester.execute(CLASS_NAME, methodDescriptor, initial);

        HeapItem item = graph.getTerminatingRegisterConsensus(0);
        Assert.assertEquals(exceptionClass, item.getValue().getClass());
        Assert.assertEquals(ClassNameUtils.toInternal(exceptionClass), item.getType());

        assertFalse("Should not reach next instruction in non-exception execution path",
                graph.wasAddressReached(2));

        MethodState mState = graph.getNodePile(0).get(0).getContext().getMethodState();
        Assert.assertEquals(0, mState.getRegistersAssigned().length);
    }

    @Test
    public void canGet() {
        initial.setRegisters(0, new int[] { 0x42 }, "[I", 1, 0, "I");
        expected.setRegisters(0, 0x42, "I");

        VMTester.test(CLASS_NAME, "get()V", initial, expected);
    }

    @Test
    public void canGetBoolean() {
        initial.setRegisters(0, new boolean[] { true }, "[Z", 1, 0, "I");
        expected.setRegisters(0, true, "Z");

        VMTester.test(CLASS_NAME, "getBoolean()V", initial, expected);
    }

    @Test
    public void canGetByte() {
        initial.setRegisters(0, new byte[] { 0xe }, "[B", 1, 0, "I");
        expected.setRegisters(0, (byte) 0xe, "B");

        VMTester.test(CLASS_NAME, "getByte()V", initial, expected);
    }

    @Test
    public void canGetChar() {
        initial.setRegisters(0, new char[] { 'a' }, "[C", 1, 0, "I");
        expected.setRegisters(0, 'a', "C");

        VMTester.test(CLASS_NAME, "getChar()V", initial, expected);
    }

    @Test
    public void canGetObject() {
        String objectValue = "stringy";
        String[] objectArray = new String[] { objectValue };
        String objectType = "Ljava/lang/String;";
        initial.setRegisters(0, objectArray, objectType, 1, 0, "I");
        expected.setRegisters(0, objectValue, objectType);

        VMTester.test(CLASS_NAME, "getObject()V", initial, expected);
    }

    @Test
    public void canGetShort() {
        initial.setRegisters(0, new short[] { 0x42 }, "[S", 1, 0, "I");
        expected.setRegisters(0, (short) 0x42, "S");

        VMTester.test(CLASS_NAME, "getShort()V", initial, expected);
    }

    @Test
    public void canGetUninitializedPrimitive() {
        initial.setRegisters(0, new int[1], "[I", 1, 0, "I");
        expected.setRegisters(0, (new int[1])[0], "I");

        VMTester.test(CLASS_NAME, "getUninitializedInt()V", initial, expected);
    }

    @Test
    public void canGetUnknownArray() {
        initial.setRegisters(0, new UnknownValue(), "[I", 1, 0, "I");
        expected.setRegisters(0, new UnknownValue(), "I");

        VMTester.test(CLASS_NAME, "get()V", initial, expected);
    }

    @Test
    public void canGetUnknownElement() {
        initial.setRegisters(0, new Object[] { new UnknownValue(), 5 }, "[I", 1, 0, "I");
        expected.setRegisters(0, new UnknownValue(), "I");

        VMTester.test(CLASS_NAME, "get()V", initial, expected);
    }

    @Test
    public void canGetUnknownIndex() {
        initial.setRegisters(0, new int[] { 0x42 }, "[I", 1, new UnknownValue(), "I");
        expected.setRegisters(0, new UnknownValue(), "I");

        VMTester.test(CLASS_NAME, "get()V", initial, expected);
    }

    @Test
    public void canGetWide() {
        initial.setRegisters(0, new long[] { 0x10000000000L }, "J", 1, 0, "I");
        expected.setRegisters(0, 0x10000000000L, "J");

        VMTester.test(CLASS_NAME, "getWide()V", initial, expected);
    }

    @Test
    public void canGetWithShortIndex() {
        Short index = 0;
        initial.setRegisters(0, new int[] { 0x42 }, "[I", 1, index, "S");
        expected.setRegisters(index.intValue(), 0x42, "S");

        VMTester.test(CLASS_NAME, "get()V", initial, expected);
    }

    @Test
    public void nullArrayValueThrowsNullPointerExceptionAndHasNoChildrenAndAssignsNoRegisters() {
        initial.setRegisters(0, null, "[I", 1, 0, "I");
        testException("getWithCatch()V", NullPointerException.class, initial);
    }

    @Test
    public void outOfBoundsIndexThrowsArrayIndexOutOfBoundsExceptionAndHasNoChildrenAndAssignsNoRegisters() {
        initial.setRegisters(0, new int[5], "[I", 1, 10, "I");
        testException("getWithCatch()V", ArrayIndexOutOfBoundsException.class, initial);
    }

    @Before
    public void setUp() {
        initial = new VMState();
        expected = new VMState();
    }

}

File: smalivm/src/test/java/org/cf/smalivm/opcode/AGetOpTest.kt
package org.cf.smalivm.opcode

import org.cf.smalivm.TestState
import org.cf.smalivm.Tester
import org.cf.smalivm.type.UnknownValue
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test

class AGetOpTest {
    private lateinit var expected: TestState
    private lateinit var initial: TestState

    companion object {
        private const val CLASS_NAME = "Laget_test;"
    }

    @BeforeEach
    fun setUp() {
        initial = TestState()
        expected = TestState()
    }

    @Test
    fun canGet() {
        initial.setRegisters(0, intArrayOf(0x42), "[I", 1, 0, "I")
        expected.setRegisters(0, 0x42, "I")
        Tester.test(CLASS_NAME, "get()V", initial, expected)
    }

    @Test
    fun canGetBoolean() {
        initial.setRegisters(0, booleanArrayOf(true), "[Z", 1, 0, "I")
        expected.setRegisters(0, true, "Z")
        Tester.test(CLASS_NAME, "getBoolean()V", initial, expected)
    }

    @Test
    fun canGetByte() {
        initial.setRegisters(0, byteArrayOf(0xe), "[B", 1, 0, "I")
        expected.setRegisters(0, 0xe.toByte(), "B")
        Tester.test(CLASS_NAME, "getByte()V", initial, expected)
    }

    @Test
    fun canGetChar() {
        initial.setRegisters(0, charArrayOf('a'), "[C", 1, 0, "I")
        expected.setRegisters(0, 'a', "C")
        Tester.test(CLASS_NAME, "getChar()V", initial, expected)
    }

    @Test
    fun canGetObject() {
        val objectValue = "stringy"
        val objectArray = arrayOf(objectValue)
        val objectType = "Ljava/lang/String;"
        initial.setRegisters(0, objectArray, objectType, 1, 0, "I")
        expected.setRegisters(0, objectValue, objectType)
        Tester.test(CLASS_NAME, "getObject()V", initial, expected)
    }

    @Test
    fun canGetShort() {
        initial.setRegisters(0, shortArrayOf(0x42), "[S", 1, 0, "I")
        expected.setRegisters(0, 0x42.toShort(), "S")
        Tester.test(CLASS_NAME, "getShort()V", initial, expected)
    }

    @Test
    fun canGetUninitializedPrimitive() {
        initial.setRegisters(0, IntArray(1), "[I", 1, 0, "I")
        expected.setRegisters(0, IntArray(1)[0], "I")
        Tester.test(CLASS_NAME, "getUninitializedInt()V", initial, expected)
    }

    @Test
    fun canGetUnknownArray() {
        initial.setRegisters(0, UnknownValue(), "[I", 1, 0, "I")
        expected.setRegisters(0, UnknownValue(), "I")
        Tester.test(CLASS_NAME, "get()V", initial, expected)
    }

    @Test
    fun canGetUnknownElement() {
        initial.setRegisters(0, arrayOf(UnknownValue(), 5), "[I", 1, 0, "I")
        expected.setRegisters(0, UnknownValue(), "I")
        Tester.test(CLASS_NAME, "get()V", initial, expected)
    }

    @Test
    fun canGetUnknownIndex() {
        initial.setRegisters(0, intArrayOf(0x42), "[I", 1, UnknownValue(), "I")
        expected.setRegisters(0, UnknownValue(), "I")
        Tester.test(CLASS_NAME, "get()V", initial, expected)
    }

    @Test
    fun canGetWide() {
        initial.setRegisters(0, longArrayOf(0x10000000000L), "J", 1, 0, "I")
        expected.setRegisters(0, 0x10000000000L, "J")
        Tester.test(CLASS_NAME, "getWide()V", initial, expected)
    }

    @Test
    fun canGetWithShortIndex() {
        val index: Short = 0
        initial.setRegisters(0, intArrayOf(0x42), "[I", 1, index, "S")
        expected.setRegisters(index.toInt(), 0x42, "I", 1, index, "S")
        Tester.test(CLASS_NAME, "get()V", initial, expected)
    }

    @Test
    fun nullArrayValueThrowsNullPointerExceptionAndHasNoChildrenAndAssignsNoRegisters() {
        initial.setRegisters(0, null, "[I", 1, 0, "I")
        Tester.testSimpleException(CLASS_NAME, "getWithCatch()V", NullPointerException::class.java, initial, 2)
    }

    @Test
    fun outOfBoundsIndexThrowsArrayIndexOutOfBoundsExceptionAndHasNoChildrenAndAssignsNoRegisters() {
        initial.setRegisters(0, IntArray(5), "[I", 1, 10, "I")
        Tester.testSimpleException(CLASS_NAME, "getWithCatch()V", ArrayIndexOutOfBoundsException::class.java, initial, 2)
    }
}
